File: moderate/reverse_groups.pl
use strict;
open( INFILE, "<", $ARGV[0] )
  or die("Cannot open file $ARGV[0] for reading: $!");
while ( my $line = <INFILE> ) {
    chomp $line;
    my @s = split( /;/, $line );
    my $n = pop @s;
    my @t = split( /,/, $s[0] );
    for ( my $c = $n ; $c <= scalar @t ; $c += $n ) {
        for ( my $i = 1 ; 2 * $i <= $n ; $i++ ) {
            ( $t[ $c - $i ], $t[ $c - $n + $i - 1 ] ) =
              ( $t[ $c - $n + $i - 1 ], $t[ $c - $i ] );
        }
    }
    printf( "%s\n", join( ',', @t ) );
}
close(INFILE);

File: moderate/reverse_groups.rb
File.open(ARGV[0]).each_line do |line|
  s = line.split(';')
  t = s[0].split(',')
  n = s[1].to_i
  n.step(t.length, n) do |c|
    (1..n / 2).each do |i|
      t[c - i], t[c - n + i - 1] = t[c - n + i - 1], t[c - i]
    end
  end
  puts t.join(',')
end
